b2sdk/replication/monitoring.py

```diff
@@ -0,0 +1,281 @@
+    """
+
+    @abstractmethod
+    def add(self, source_file: Optional[B2Path], destination_file: Optional[B2Path]):
+        pass
+
+
+@dataclass
+class CountAndSampleReplicationReport(AbstractReplicationReport):
+    """
+    Replication report which groups and counts files by their `ReplicationAttrs` and
+    also stores first and last seen examples of such files.
+    """
+    counter_by_status: Counter[ReplicationAttrs] = field(default_factory=Counter)
+    samples_by_status_first: Dict[ReplicationAttrs, Tuple[FileVersion, FileVersion]] = field(
+        default_factory=dict
+    )
+    samples_by_status_last: Dict[ReplicationAttrs, Tuple[FileVersion, FileVersion]] = field(
+        default_factory=dict
+    )
+
+    def add(self, source_file: Optional[B2Path], destination_file: Optional[B2Path] = None):
+        status = ReplicationAttrs.from_files(source_file, destination_file)
+        self.counter_by_status[status] += 1
+
+        sample = (
+            source_file and source_file.selected_version,
+            destination_file and destination_file.selected_version,
+        )
+        self.samples_by_status_first.setdefault(status, sample)
+        self.samples_by_status_last[status] = sample
+
+
+@dataclass
+class ReplicationMonitor:
+    """
+    Calculates source and (optionally) destination replication statistics.
+
+    :param b2sdk.v2.Bucket bucket: replication source bucket
+    :param b2sdk.v2.ReplicationRule rule: replication rule to be monitored;
+    should belong to `bucket`'s replication configuration
+    :param b2sdk.v2.B2Api destination_api: B2Api instance for destination
+    bucket; if destination bucket is on the same account as source bucket,
+    omit this parameter and then source bucket's B2Api will be used
+    :param type replication_report_class: subclass of ReplicationReport,
+    used to aggregate files to statistics
+    :param b2sdk.v2.Report report: instance of Report which will report
+    scanning progress, by default to stdout
+    :param b2sdk.v2.ScanPoliciesManager scan_policies_manager: a strategy to scan
+    files, so that several files that match some criteria may be omitted
+    :rtype: b2sdk.v2.ReplicationMonitor
+    """
+
+    bucket: Bucket
+    rule: ReplicationRule
+    destination_api: Optional[B2Api] = None  # if None -> will use `api` of source (bucket)
+    replication_report_class: Type[AbstractReplicationReport] = CountAndSampleReplicationReport
+    report: Report = field(default_factory=lambda: Report(sys.stdout, False))
+    scan_policies_manager: ScanPoliciesManager = DEFAULT_SCAN_MANAGER
+
+    B2_FOLDER_CLASS: ClassVar[Type] = B2Folder
+    QUEUE_SIZE: ClassVar[int] = 100
+
+    def __post_init__(self):
+        if not self.bucket.replication_configuration:
+            raise ValueError(f'Bucket {self.bucket} has no replication configuration')
+
+        if self.rule not in self.bucket.replication_configuration.rules:
```

Comment:
do we have a comparator there? Because I think this is using object identity by default which is not good enough because you can have two different objects representing the same replication rule, just created twice

Reply:
It's auto-done for dataclasses:
> eq: If true (the default), an __eq__() method will be generated. This method compares the class as if it were a tuple of its fields, in order. Both instances in the comparison must be of the identical type.